### ISSUE TITLE
problem when I use DistributedAdasumOptimizer

### ISSUE THREAD

**Issue by canisn** [horovod/horovod#1729]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)
TensorFlow
2. Framework version:
1.14.0
3. Horovod version:
1.19.0
4. MPI version:
4.0.2
5. CUDA version:
no
6. NCCL version:
no
7. Python version:
3.6.5
8. OS and version:
Linux version 3.10.0-862.3.3.el7.x86_64 (builder@kbuilder.dev.centos.org) 
9. GCC version:
(gcc version 4.8.5 20150623 (Red Hat 4.8.5-28) (GCC) )
**Checklist:**
1. Did you search issues to find if somebody asked this question before?
yes
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Bug report:**
Please describe errorneous behavior you're observing and steps to reproduce it.
When I set backward_passes_per_step greater than 1, I was prompted to set op = adasum. But in this case, the accuracy is worse than a single machine. Looking at the source code, in the class _DistributedAdasumOptimizer, the compute_gradients and apply_gradients methods are methods that return directly to the called stand-alone optimizer. They do not use the _apply_shared method in the current class. After trying to modify the calling relationship here, I found this sentence, there is a bug in start_slot = self._get_or_make_slot (var, "delta_start").
Is there a problem here? Thank you.

**Comment by abditag2** [comment]
We do not call the stand-alone optimizer, rather we dynamically create a new python optimizer class, with the same name and attributes of the original optimizer except we overwrite get_gradients with all_reduced gradients. That is why if you print the type of the distributed_optimizer, you will get the same name as the original optimizer that you passed in. If you inspect more with `ipdb` you can see that the overridden get_gradients method.

**Comment by zmldndx** [comment]
I got the same problem. _DistributeAdasumOptimizer does not call _apply_shared which calls allreduce(In _apply_shared function, _get_or_make_slot with 2 inputs is not found in source code), it seems that it only compute and apply gradients locally. I changed __init__.py under horovod/tensorflow as follow, please help to check if there is any problem:
https://github.com/horovod/horovod/compare/master...zmldndx:patch-1

**Comment by canisn** [comment]
@abditag2  Thanks for your reply.
However, I still think that when the adasum strategy is executed, all_reduce operations are not performed between multiple machines. There is evidence to prove my point. When I use distributed training for a period of time, I save checkpoints under each machine. These checkpoints are used to predict the same data, and their predictions vary widely.
In addition, the same problem as @zmldndx, the definition of the function self._get_or_make_slot (var, "delta_start") was not found throughout the project, with two parameters. There is a four parameter in tf.

**Comment by canisn** [comment]
@abditag2 ,thank you for your reply.
However, I still think that when the adasum strategy is executed, all_reduce operations are not performed between multiple machines. There is evidence to prove my point. When I use distributed training for a period of time, I save checkpoints under each machine. These checkpoints are used to predict the same data, and their predictions vary widely.
In addition, the same problem as @zmldndx, the definition of the function get_or_make_slot(var, "delta_start"),with two parameters, was not found throughout the project.

**Comment by tgaddair** [comment]
@Tixxx @kit1980 could you take a look at this issue?

**Comment by PANXiao1994** [comment]
> @abditag2 ,thank you for your reply.
> However, I still think that when the adasum strategy is executed, all_reduce operations are not performed between multiple machines. There is evidence to prove my point. When I use distributed training for a period of time, I save checkpoints under each machine. These checkpoints are used to predict the same data, and their predictions vary widely.
> In addition, the same problem as @zmldndx, the definition of the function get_or_make_slot(var, "delta_start"),with two parameters, was not found throughout the project.

Same problem here. I encountered the same problem a couple of weeks ago. My assumption is the same as yours. [https://github.com/horovod/horovod/issues/1676](url) and [https://github.com/horovod/horovod/issues/1667](url)
